### PR TITLE
chore: refactoring amplitude using optional chaining

### PR DESCRIPTION
**Pull request by anantjain45823**
## Description of the change

> Resolves INT-152
Used Optional Chaining 
Increased test coverage
Added Validations 
## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/util/index.js [hunk 1]
```diff
@@ -2039,7 +2039,14 @@ const getAuthErrCategoryFromStCode = (status) => {
...
   return '';
 };
-
+const isValidInteger = (value) => {
+  if (Number.isNaN(value) || !isDefinedAndNotNull(value)) {
+    return false;
+  }
+  // const valueToString = `${value}`;
+  // Use a regular expression to check if the string is a valid integer or a valid floating-point number
+  return /^-?\d+(\.\d+)?$/.test(`${value}`);
```
Comment (koladilip) [inline src/v0/util/index.js:2048]: should we try to do this after typeof value === 'string'

Comment (anantjain45823) [reply src/v0/util/index.js]: done

File: src/v0/util/index.js [hunk 2]
```diff
@@ -2052,8 +2052,13 @@ const isValidInteger = (value) => {
   }
   // const valueToString = `${value}`;
   // Use a regular expression to check if the string is a valid integer or a valid floating-point number
-  return /^-?\d+(\.\d+)?$/.test(`${value}`);
+  return typeof value === 'string' ? /^-?\d+(\.\d+)?$/.test(`${value}`) : false;
```
Comment (koladilip) [inline src/v0/util/index.js:2055]: typeof value === 'string' ? /^-?\d+(\.\d+)?$/.test(value) : false;

**Comment by Gauravudia** [comment]
@anantjain45823 Add unit tescases of your changes

**Comment by anantjain45823** [comment]
> @anantjain45823 Add unit test cases of your changes
There is not anything major done but big chunk of functions divided into small ones.

File: src/v0/util/index.js [hunk 3]
```diff
@@ -2052,17 +2052,24 @@ const getAuthErrCategoryFromStCode = (status) => {
...
+
 const isValidInteger = (value) => {
   if (Number.isNaN(value) || !isDefinedAndNotNull(value)) {
     return false;
   }
-  // const valueToString = `${value}`;
+  const valueToString = `${value}`;
   // Use a regular expression to check if the string is a valid integer or a valid floating-point number
-  return typeof value === 'string' ? /^-?\d+(\.\d+)?$/.test(`${value}`) : false;
+  return /^-?\d+$/.test(valueToString);
```
Comment (koladilip) [inline src/v0/util/index.js:2062]: why did you remove typeof checking?

Comment (koladilip) [reply src/v0/util/index.js:2062]:
```js
if (typeof value === 'number' ) {
    return true;
} else if (typeof value ===  'string') {
   return /^-?\d+$/.test(value);
} else {
  return false;
}
```

File: src/v0/util/index.js [hunk 4]
```diff
@@ -2052,17 +2052,24 @@ const getAuthErrCategoryFromStCode = (status) => {
...
+};
+const validateEventType = (event) => {
```
Comment (koladilip) [inline src/v0/util/index.js:2064]: seems like repeated

File: src/v0/util/index.js (hunk 3 above)
Comment (anantjain45823) [reply src/v0/util/index.js:2062]: we are converting the input value to string and then checking it through regex expression to avoid multiple type checks

File: src/v0/util/index.js (hunk 4 above)
Comment (anantjain45823) [reply src/v0/util/index.js:2064]: we are converting the input value to string and then checking it through regex expression to avoid multiple type checks

File: src/v0/util/index.js (hunk 3 above)
Comment (koladilip) [reply src/v0/util/index.js:2062]: won't that be slow?

File: src/v0/destinations/am/transform.js [hunk 5]
```diff
@@ -79,112 +80,129 @@ const aliasEndpoint = (destConfig) => {
...
-function prepareTraitsConfig(configPropertyTrait, actionKey, traitsObject) {
+const prepareTraitsConfig = (configPropertyTrait, actionKey, traitsObject) => {
   traitsObject[actionKey] = {};
   configPropertyTrait.forEach((traitsElement) => {
-    const property = traitsElement.traits;
+    const property = traitsElement?.traits;
     traitsObject = updateTraitsObject(property, traitsObject, actionKey);
   });
-  if (Object.keys(traitsObject[actionKey]).length === 0) {
+  if (Object.keys(traitsObject[actionKey])?.length === 0) {
```
Comment (sanpj2292) [inline src/v0/destinations/am/transform.js:158]:
```suggestion
  if (typeof traitsObject?.[actionKey] === 'object' && Object.keys(traitsObject?.[actionKey] || {})?.length === 0) {
```

File: src/v0/destinations/am/transform.js [hunk 6]
```diff
@@ -79,112 +80,129 @@ const aliasEndpoint = (destConfig) => {
...
-  rawPayload.revenue = message.properties.revenue;
+  rawPayload.revenue = message.properties?.revenue;
   return rawPayload;
-}
+};
 
-function createRevenuePayload(message, rawPayload) {
-  rawPayload.productId = message.properties.product_id;
+const createRevenuePayload = (message, rawPayload) => {
+  rawPayload.productId = message?.properties?.product_id;
```
Comment (sanpj2292) [inline src/v0/destinations/am/transform.js:136]:
```suggestion
  rawPayload.productId = message.properties?.product_id;
```

File: src/v0/util/index.js (hunk 3 above)
Comment (anantjain45823) [reply src/v0/util/index.js:2062]: It's a trade-off between multiple checks for data type or convert to string and then use regex .
What's your opinion

Comment (koladilip) [reply src/v0/util/index.js:2062]: it is just string or number check so it is not that complex in my opinion. I avoid converting to string as it will be almost costly

Comment (anantjain45823) [reply src/v0/util/index.js:2062]: Cool will do it in that way then

Comment (anantjain45823) [reply src/v0/util/index.js]: Done

**Review by mihir-4116 — DISMISSED** [review]

**Review by koladilip — DISMISSED** [review]

File: src/v0/destinations/am/transform.js [hunk 7]
```diff
@@ -419,113 +541,58 @@ function responseBuilderSimple(
...
   }
-
   return respList;
-}
+};
 
+const getGroupInfo = (destination, groupInfo, groupTraits) => {
+  let updatedGroupInfo = { ...groupInfo };
+  const groupTypeTrait = get(destination, 'Config.groupTypeTrait');
+  const groupValueTrait = get(destination, 'Config.groupValueTrait');
```
Comment (sanpj2292) [inline src/v0/destinations/am/transform.js:559]:
```suggestion
  const { groupTypeTrait, groupValueTrait } = destination.Config;
```

Comment (sanpj2292) [reply src/v0/destinations/am/transform.js:559]: Also I think we can move `let updatedGroupInfo` into the if-block

File: src/v0/destinations/am/transform.js [hunk 8]
```diff
@@ -682,16 +726,16 @@ function getProductPurchasedEvents(message, destination) {
     });
   }
   return productPurchasedEvents;
-}
+};
 
-function trackRevenueEvent(message, destination) {
+const trackRevenueEvent = (message, destination) => {
   let sendEvents = [];
   const originalEvent = JSON.parse(JSON.stringify(message));
```
Comment (sanpj2292) [inline src/v0/destinations/am/transform.js:733]:
```suggestion
  const originalEvent = cloneDeep(message);
```

File: src/v0/destinations/am/transform.js [hunk 9]
```diff
@@ -782,7 +830,7 @@ function getBatchEvents(message, destination, metadata, batchEventResponse) {
   if (batchEventArray.length === 0) {
     if (JSON.stringify(incomingMessageJSON).length < AMBatchSizeLimit) {
       delete message.body.JSON.options;
-      batchEventResponse = Object.assign(batchEventResponse, {
+      Object.assign(batchEventResponse, {
```
Comment (sanpj2292) [inline src/v0/destinations/am/transform.js:833]: Can we do this ?
```js
batchEventResponse.batchedRequest = message;
```

**Review by sanpj2292 — CHANGES_REQUESTED** [review]

File: src/v0/destinations/am/transform.js [hunk 10]
```diff
@@ -303,50 +346,129 @@ function getResponseData(evType, destination, rawPayload, message, groupInfo) {
...
+    buildPayloadForMobileChannel(message, destination, payload);
+  }
+  payload.time = new Date(getFieldValueFromMessage(message, 'timestamp')).getTime();
+
+  // send user_id only when present, for anonymous users not required
+  if (
+    message.userId &&
+    message.userId !== '' &&
+    message.userId !== 'null' &&
+    message.userId !== null
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/am/transform.js:396]: It would definitely

File: src/v0/destinations/am/transform.js [hunk 11]
```diff
@@ -79,112 +80,129 @@ const aliasEndpoint = (destConfig) => {
...
-  return get(message, 'session_id')
-    ? handleSessionIdUnderRoot(message)
-    : get(message, 'context.sessionId')
-    ? handleSessionIdUnderContext(message)
-    : -1;
-}
+const checkForJSONAndUserIdLengthAndDeviceId = (jsonBody, userId, deviceId) =>
+  Object.keys(jsonBody).length === 0 ||
+  (!batchEventsWithUserIdLengthLowerThanFive && userId && userId.length < 5) ||
+  (batchEventsWithUserIdLengthLowerThanFive && userId && userId.length < 5 && !deviceId);
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/am/transform.js]: Done

File: src/v0/destinations/am/transform.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/am/transform.js]: Changed

File: src/v0/destinations/am/transform.js [hunk 12]
```diff
@@ -79,112 +80,137 @@ const aliasEndpoint = (destConfig) => {
...
     traitsObject = prepareTraitsConfig(destination.Config.traitsToPrepend, actionKey, traitsObject);
   }
   return traitsObject;
-}
+};
+
+const getScreenevTypeAndUpdatedProperties = (message, CATEGORY_KEY) => {
+  const name = message.name || message.event || get(message, CATEGORY_KEY);
+  return {
+    evType: `Viewed ${name || ''} Screen`,
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/am/transform.js]: Not sure if its or not but it would look bad so changes this with test case but I have seen other instances of this as well

File: src/v0/destinations/am/transform.js (hunk 7 above)
Comment (anantjain45823) [reply src/v0/destinations/am/transform.js]: Done

File: src/v0/destinations/am/transform.js (hunk 8 above)
Comment (anantjain45823) [reply src/v0/destinations/am/transform.js]: Done

File: src/v0/destinations/am/transform.js (hunk 9 above)
Comment (anantjain45823) [reply src/v0/destinations/am/transform.js]: Done

File: src/v0/destinations/am/transform.js [hunk 13]
```diff
@@ -391,7 +513,8 @@ function responseBuilderSimple(
 
   const respData = getResponseData(evType, destination, rawPayload, message, groupInfo);
   const { groups } = respData;
-  ({ endpoint, rawPayload } = respData);
+  // eslint-disable-next-line prefer-const
+  ({ rawPayload } = respData);
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/am/transform.js]: NA anymore

File: src/v0/destinations/am/transform.js [hunk 14]
```diff
@@ -243,57 +261,82 @@ function updateConfigProperty(message, payload, mappingJson, validatePayload, Co
...
+    rawPayload = createRevenuePayload(message, rawPayload);
+    // deleting the properties price, product_id, quantity and revenue from event_properties since it is already in root
+    if (rawPayload.event_properties) {
+      delete rawPayload.event_properties.price;
+      delete rawPayload.event_properties.product_id;
+      delete rawPayload.event_properties.quantity;
+      delete rawPayload.event_properties.revenue;
+    }
+  }
+  const groups = groupInfo && Object.assign(groupInfo);
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/am/transform.js]: Done

File: src/v0/destinations/am/transform.js [hunk 15]
```diff
@@ -79,112 +80,138 @@ const aliasEndpoint = (destConfig) => {
...
+    quantity = message.properties?.quantity || 1;
   } else {
-    price = message.properties.revenue;
+    price = message.properties?.revenue;
     quantity = 1;
   }
   rawPayload.price = price;
   rawPayload.quantity = quantity;
-  rawPayload.revenue = message.properties.revenue;
+  rawPayload.revenue = message.properties?.revenue;
```
Comment (sanpj2292) [inline src/v0/destinations/am/transform.js:133]: [Q] If `properties.revenue` is not present, `revenue` value would be `undefined`, will it rightly put data into Amplitude ? 🤔

Comment (sanpj2292) [reply src/v0/destinations/am/transform.js:133]: Should it default to valid value which indicates nothing like may be `0` or `0.0` ?

File: src/v0/destinations/am/transform.js [hunk 16]
```diff
@@ -545,26 +607,26 @@ function processSingleMessage(message, destination) {
...
         evType =
           userProvidedPageEventString.trim() === ''
-            ? message.name
+            ? name
             : userProvidedPageEventString
                 .trim()
                 .replaceAll(/{{([^{}]+)}}/g, get(message, getMessagePath));
       } else {
-        evType = `Viewed ${message.name || get(message, CATEGORY_KEY) || ''} Page`;
+        evType = `Viewed ${name || get(message, CATEGORY_KEY) || ''} Page`;
```
Comment (sanpj2292) [inline src/v0/destinations/am/transform.js:615]: Spacing concern

File: src/v0/destinations/am/transform.js (hunk 15 above)
Comment (anantjain45823) [reply src/v0/destinations/am/transform.js:133]: Yess we won't be calculating revenue from our side for this one as already discussed its a customer expected property and they can easily use user transformation for this.
Reasons we are not doing it at our end:
1. Not much destinations do this
2. breaking change
3. Easily achievable through ut

File: src/v0/destinations/am/transform.js (hunk 16 above)
Comment (anantjain45823) [reply src/v0/destinations/am/transform.js]: done

**Review by sanpj2292 — APPROVED** [review]

**Review by sanpj2292 — APPROVED** [review]

**Review by mihir-4116 — APPROVED** [review]
